Caas

cadecairos opened the pull request:
Related issue: https://github.com/mozilla/foundation.mozilla.org/issues/1802
Summary: Implement Product voting for PNI, implement a script to periodically summarize votes for each product, implement a serializer that outputs product vote statistics.

cadecairos commented:
@gvn I wrote a Serializer that generates the JSON as we described in #1802, and use it to generate the objects passed to the views here: https://github.com/mozilla/foundation.mozilla.org/pull/1864/files#diff-fdf5c15be366077e370c1d3b6d42def7R24

We also needed ways to query/filter products, right? Do we have details on what exactly is required?

patjouk commented:
Cool stuff! Can you add tests around the management command to check if the tally of the vote breakdown looks like we expect it to be? Thanks!

patjouk commented on network-api/networkapi/buyersguide/models.py:
```diff
@@ -148,3 +154,125 @@ class Product(models.Model):
+
+class ProductVote(models.Model):
+    votes = models.IntegerField(
+        null=False,
```
It's at [False by default](https://github.com/django/django/blob/stable/1.11.x/django/db/models/fields/__init__.py#L145-L151), you don't need to specify it.

patjouk commented on network-api/networkapi/buyersguide/models.py:
```diff
@@ -148,3 +154,125 @@ class Product(models.Model):
+class RangeProductVote(ProductVote):
+    attribute = models.CharField(
+        max_length=100,
+        null=False,
```
same as before

patjouk commented on network-api/networkapi/buyersguide/models.py:
```diff
@@ -148,3 +154,125 @@ class Product(models.Model):
+        ]
+    )
+    average = models.IntegerField(
+        null=False,
```
I won't point all of them, but you get the idea ;)

gvn commented on network-api/networkapi/buyersguide/serializers.py:
```diff
@@ -0,0 +1,56 @@
+
+            votes.append(breakdown)
+
+        return {
```
The schema has shifted a little bit for Product in my latest PR:

https://github.com/mozilla/foundation.mozilla.org/blob/c704458880858e4da1478bc1f1172868be9463c0/network-api/networkapi/buyersguide/models.py#L42

cadecairos replied:
Should I base my PR on yours?

gvn replied:
Yeah, it's actually landed now.

gideonthomas commented on network-api/networkapi/buyersguide/serializers.py:
```diff
@@ -0,0 +1,56 @@
+            votes.append(breakdown)
+
+        return {
+            'name': instance.name,
```
nit: use the `model_to_dict` helper to generate the dict: https://docs.djangoproject.com/en/2.1/_modules/django/forms/models/

gideonthomas commented on network-api/networkapi/buyersguide/views.py:
```diff
@@ -1,23 +1,84 @@
+        return Response('Invalid product', status=400, content_type='text/plain')
+    except ValidationError as ex:
+        return Response(f'Payload validation failed: {ex}', status=400, content_type='text/plain')
+    except Error:
```
let's at least log this

cadecairos commented on network-api/networkapi/buyersguide/views.py:
```diff
@@ -1,23 +1,84 @@
 def buyersguide_home(request):
     products = Product.objects.all()
-    return render(request, 'buyersguide_home.html', {'products': products})
+    serialized_products = [ProductSerializer(product) for product in products]
```
well that's handy!

cadecairos commented on network-api/networkapi/buyersguide/views.py:
```diff
@@ -1,23 +1,84 @@
     products = Product.objects.all()
-    return render(request, 'buyersguide_home.html', {'products': products})
+    serialized_products = [ProductSerializer(product) for product in products]
+    return render(request, 'buyersguide_home.html', {'products': serialized_products})
```
In this case the dict being passed here is the context when rendering the template. (I think)

gideonthomas replied:
is there a reason we're serializing it then? As opposed to just passing in the model instances directly?

cadecairos replied:
We're using the serializer to generate a `votes` attribute that presents the aggregate vote data in a nice format.

gideonthomas replied:
serializer seems to be extra overhead for just that one thing though. Are you able to just do it in this function itself since you need the rest of the model data anyway?

cadecairos replied:
Maybe, I'll look into an alternative implementation

kristinashu commented:
cc @Pomax

cadecairos commented:
@patjouk I've added a test that checks that the management script generates the right vote counts

@gideonthomas I've removed that Serializer and set up the `votes` attribute on the model directly.

@gvn I'm still not sure what API routes we might need here. I'm passing the new attribute to the existing templates but do we need to expose products with an API endpoint?

gvn commented:
Yeah, we'll need to expose products with an route as well. I wasn't sure if you wanted to do that in this PR or later, but feel free to do either approach.

cadecairos commented:
Let's land this and get a DRF route in another ticket

Pomax commented on network-api/networkapi/buyersguide/models.py:
```diff
@@ -1,6 +1,14 @@
 from networkapi.utility.images import get_image_upload_path
-# from wagtail.snippets.models import register_snippet
+
+VALID_VOTE_ATTRIBUTES = [
```
crufty?

Pomax commented on network-api/networkapi/buyersguide/views.py:
```diff
@@ -1,23 +1,84 @@
 def product_view(request, productname):
     product = Product.objects.get(name__iexact=productname)
-    return render(request, 'product_page.html', {'product': product, 'mediaUrl': settings.MEDIA_URL})
+    return render(request, 'product_page.html', {'product': product.to_dict()})
```
we probably want to keep that mediaURL, and then figure out of we still use it in a new issue

Pomax commented on network-api/networkapi/buyersguide/views.py:
```diff
@@ -1,23 +1,84 @@
+
+    try:
+        product = Product.objects.get(id=product_id)
+        vote_class = RangeVote
```
can we capitalise this as a class? (VoteClass maybe?)

Pomax commented on network-api/networkapi/settings.py:
```diff
@@ -487,3 +489,9 @@ def show_toolbar(request):
+BUYERS_GUIDE_VOTE_RATE_LIMIT = env('BUYERS_GUIDE_VOTE_RATE_LIMIT')
+
+# Detect if we're testing
+TESTING = sys.argv[1:2] == ['test']
```
this feels like it's position based, probably a good idea to harden this by doing an `in` check, like `TESTING = 'test' in sys.argv`?

Pomax commented on network-api/networkapi/buyersguide/management/commands/aggregate_product_votes.py:
```diff
@@ -0,0 +1,118 @@
+            creepiness_bucket_totals = {}
+
+            # For each bucket group, filter votes on the low and high values, and record the count of each
+            for creepiness_bucket in self.range_filter_values.keys():
```
what is `self` here?

cadecairos replied:
OH. `self` is the Test instance. it's referring to this variable: https://github.com/mozilla/foundation.mozilla.org/pull/1864/files/1f4ffa989c910c3d2cbbce049f48bc895a65e57e#diff-faeee7b1bd84dcf1f1ecd9d0661bb9b9R21

Pomax replied:
404 on that link?

(we might want to add a comment so we keep that knowledge local to the code =D)

cadecairos replied:
yeah I rebased the branch so the hashes changed. I think it's clear if you look at parameters for the method this code is in (something I missed while skimming through the code.)

Pomax commented on network-api/networkapi/buyersguide/tests.py:
```diff
@@ -1,3 +1,272 @@
+class ManagementCommandTest(APITestCase):
+
+    def setUp(self):
+        self.vote_url = reverse('product-vote')
```
nit, but: this doesn't need to be here, we can get the `reverse('product-vote') value outside the class, as a constant, and refer to that in both `ManagementCommandTest` and `BuyersGuideVoteTest`

Pomax commented on network-api/networkapi/buyersguide/views.py:
```diff
@@ -1,23 +1,88 @@
 @login_required
 def buyersguide_home(request):
     products = Product.objects.all()
+    products = [p.to_dict() for p in products]
```
just to prevent weirdness, `products = [p.to_dict() for p in Product.objects.all()]` and then just remove L22 above it?

Pomax commented on network-api/networkapi/buyersguide/views.py:
```diff
@@ -1,23 +1,88 @@
 
 @login_required
 def product_view(request, productname):
+    print(productname)
```
the dreaded print statement!

cadecairos replied:
just testing you.

cadecairos commented:
@Pomax this is ready for a re-review
